resources/assets/lib/beatmap-discussions/discussions.tsx

```diff
@@ -0,0 +1,301 @@
+  private get currentSort() {
+    return this.sort[this.props.mode];
+  }
+
+  @computed
+  private get isTimelineVisible() {
+    return this.props.mode === 'timeline' && this.currentSort === 'timeline';
+  }
+
+  @computed
+  private get sortedDiscussions() {
+    return this.props.currentDiscussions[this.props.mode].slice().sort((a: BeatmapsetDiscussionJson, b: BeatmapsetDiscussionJson) => {
+      const mapperNoteCompare =
+        // no sticky for timeline sort
+        this.currentSort !== 'timeline'
+        // stick the mapper note
+        && [a.message_type, b.message_type].includes('mapper_note')
+        // but if both are mapper note, do base comparison
+        && a.message_type !== b.message_type;
+
+      if (mapperNoteCompare) {
+        return a.message_type === 'mapper_note' ? -1 : 1;
+      } else {
+        return sortPresets[this.currentSort].sort(a, b);
+      }
+    });
+  }
+
+  constructor(props: Props) {
+    super(props);
+    makeObservable(this);
+  }
+
+  componentDidMount() {
+    $.subscribe('beatmapset-discussions:highlight', this.handleSetHighlight);
+  }
+
+  componentWillUnmount() {
+    $.unsubscribe('beatmapset-discussions:highlight', this.handleSetHighlight);
+  }
+
+  render() {
+    return (
+      <div className='osu-page osu-page--small osu-page--full'>
+        <div className={bn}>
+          <div className='page-title'>
+            {trans('beatmaps.discussions.title')}
+          </div>
+          <div className={`${bn}__toolbar`}>
+            <div className={`${bn}__toolbar-content ${bn}__toolbar-content--left`}>
+              <div className={`${bn}__toolbar-item`}>
+                {this.renderSortOptions()}
+              </div>
+            </div>
+            <div className={`${bn}__toolbar-content ${bn}__toolbar-content--right`}>
+              {this.renderShowDeletedToggle()}
+              {this.renderExpandCollapseAllButton('collapse')}
+              {this.renderExpandCollapseAllButton('expand')}
+            </div>
+          </div>
+
+          {this.renderDiscussions()}
+        </div>
+      </div>
+    );
+  }
+
+  @action
+  private readonly handleChangeSort = (e: React.SyntheticEvent<HTMLButtonElement>) => {
+    this.sort[this.props.mode] = e.currentTarget.dataset.sortPreset as Sort;
+  };
+
+  @action
+  private handleExpandClick = (e: React.SyntheticEvent<HTMLButtonElement>) => {
+    this.discussionsState.discussionDefaultCollapsed = e.currentTarget.dataset.type === 'collapse';
+    this.discussionsState.discussionCollapsed.clear();
+  };
+
+  @action
+  private readonly handleSetHighlight = (_event: unknown, { discussionId }: { discussionId: number }) => {
+    // TODO: update main to use context instead of publishing event.
+    this.discussionsState.highlightedDiscussionId = discussionId;
+  };
+
+  private readonly renderDiscussionPage = (discussion: BeatmapsetDiscussionJsonForShow) => {
+    if (discussion.id == null) return null; // TODO: does this still happen?
```

Comment:
this is for when deleted discussion was returned as empty array so it'd be weird if this still happens